broadinstitute/pearl
# AR-87  adding consent response capability

This adds the server-side processing for handling consent responses.  This includes the new SurveyTaskDispatcher, which processes events to detect survey tasks to assign to participants.  It also upgrades the previous ConsentTaskDispatcher to compute/update a master "consented" property on the enrollee, and generally refines the event-passing structures.

In general, the pattern I'm going for is:
1. Service processes API request
2. Service fires event
3. Dispatchers/listeners catch the event, do their things, and update the event with any modified state
4. Service returns a response including the latest data on the event.

We introduce a new type "HubResponse" that can be returned from API calls that impact participant state.  It includes the content of the response, as well as updated copies of the Enrollee and the ParticipantTask list.  This avoids an extra round trip when reloading the hub to display the participant tasks.

TO TEST:
1. nuke/reset db
2. rerun setup and ourhealth populate scripts
3. Go to `http://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants`
4. Confirm you see that 'Con Senter' has consented, while 'New Enrollee' has not
![image](https://user-images.githubusercontent.com/2800795/217076397-67e8cced-441d-4820-b392-7e641aca668d.png)

6. go to `http://sandbox.ourhealth.localhost:3001/`
7. sign in as "newbie@test.com"
8. confirm you see the hub with a consent task
9. go to that task and complete the consent form
10. confirm you are taken to an updated hub showing the new survey task, and a message about consent completion
![image](https://user-images.githubusercontent.com/2800795/217076504-9af0fb8a-ce8c-488a-9c94-5a34da2c917e.png)

11. reload `http://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants`
12. confirm that "New Enrollee" now appears as consented
